File: libs/power-log-parser/src/lib/parsers/overloaded-crystals-parser.ts
import { GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { Node, TagChange } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class OverloadedCrystalsParser implements ActionParser {
	readonly ParserName = 'OverloadedCrystalsParser';

	private GameState: GameState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, stateFacade: StateFacade) {
		this.GameState = parserState.GameState;
		this.StateFacade = stateFacade;
	}

	AppliesOnNewNode(node: Node, stateType: StateType): boolean {
		let tagChange: TagChange | null = null;
		return (
			stateType === StateType.PowerTaskList &&
			node.Type === TagChange &&
			((tagChange = node.Object as TagChange).Name === (GameTag.OVERLOAD_LOCKED as number) ||
				tagChange.Name === (GameTag.OVERLOAD_OWED as number))
		);
	}

	AppliesOnCloseNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	CreateGameEventProviderFromNew(node: Node): GameEventProvider[] | null {
		const tagChange = node.Object as TagChange;
		const entity = this.GameState.CurrentEntities.get(tagChange.Entity);
		if (entity == null) {
			return null;
		}

		const overloaded =
			tagChange.Name === (GameTag.OVERLOAD_LOCKED as number)
				? tagChange.Value + entity.GetTag(GameTag.OVERLOAD_OWED, 0)
				: tagChange.Value + entity.GetTag(GameTag.OVERLOAD_LOCKED, 0);
		const cardId = entity.CardId;
		const controllerId = entity.GetEffectiveController();
		return [
			GameEventProvider.Create(
				tagChange.TimeStamp,
				'OVERLOADED_CRYSTALS_CHANGED',
				GameEventHelper.CreateProvider('OVERLOADED_CRYSTALS_CHANGED', cardId, controllerId, entity.Id, this.StateFacade, {
					Overload: overloaded,
				}),
				true,
				node,
			),
		];
	}

	CreateGameEventProviderFromClose(_node: Node): GameEventProvider[] | null {
		return null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/OverloadedCrystalsParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;

namespace HearthstoneReplays.Events.Parsers
{
    public class OverloadedCrystalsParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public OverloadedCrystalsParser(ParserState ParserState, StateFacade facade)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = facade;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            TagChange tagChange = null;
            return stateType == StateType.PowerTaskList
                && node.Type == typeof(TagChange)
                && ((tagChange = node.Object as TagChange).Name == (int)GameTag.OVERLOAD_LOCKED
                    || tagChange.Name == (int)GameTag.OVERLOAD_OWED);
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return false;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            var tagChange = node.Object as TagChange;
            var entity = GameState.CurrentEntities.GetValueOrDefault(tagChange.Entity);
            if (entity == null)
            {
                return null;
            }

            var overloaded = tagChange.Name == (int)GameTag.OVERLOAD_LOCKED
                ? tagChange.Value + entity.GetTag(GameTag.OVERLOAD_OWED, 0)
                : tagChange.Value + entity.GetTag(GameTag.OVERLOAD_LOCKED, 0);
            var cardId = entity.CardId;
            var controllerId = entity.GetEffectiveController();
            return new List<GameEventProvider> { GameEventProvider.Create(
                tagChange.TimeStamp,
                 "OVERLOADED_CRYSTALS_CHANGED",
                GameEvent.CreateProvider(
                    "OVERLOADED_CRYSTALS_CHANGED",
                    cardId,
                    controllerId,
                    entity.Id,
                    StateFacade,
                    //null,
                    new {
                        Overload = overloaded,
                    }),
                true,
                node) };
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            return null;
        }
    }
}
